docs/design.md

```diff
@@ -0,0 +1,785 @@
+* `replace` - overwrite data at the specified path and replace it with the data
+  given in this document
+* `delete` - remove the data at the specified path
+
+After actions are applied for a given layer, substitutions are applied (see
+the Substitution section for details).
+
+Selection of document parents is controlled by the `parentSelector` field and
+works as follows. A given document, `C`, that specifies a `parentSelector`
+will have exactly one parent, `P`. Document `P` will be the highest
+precedence (i.e. part of the lowest layer) document that has the labels
+indicated by the `parentSelector` (and possibly additional labels) from the
+set of all documents of the same `schema` as `C` that are in layers above the
+layer `C` is in. For example, consider the following sample documents:
+
+```yaml
+---
+schema: deckhand/LayeringPolicy/v1
+metadata:
+  schema: metadata/Control/v1
+  name: layering-policy
+data:
+  layerOrder:
+    - global
+    - region
+    - site
+---
+schema: example/Kind/v1
+metadata:
+  schema: metadata/Document/v1
+  name: global-1234
+  labels:
+    key1: value1
+  layeringDefinition:
+    abstract: true
+    layer: global
+data:
+  a:
+    x: 1
+    y: 2
+---
+schema: example/Kind/v1
+metadata:
+  schema: metadata/Document/v1
+  name: region-1234
+  labels:
+    key1: value1
+  layeringDefinition:
+    abstract: true
+    layer: region
+    parentSelector:
+      key1: value1
+    actions:
+      - method: replace
+        path: .a
+data:
+  a:
+    z: 3
+---
+schema: example/Kind/v1
+metadata:
+  schema: metadata/Document/v1
+  name: site-1234
+  layeringDefinition:
+    layer: site
+    parentSelector:
+      key1: value1
+    actions:
+      - method: merge
+        path: .
+data:
+  b: 4
+...
+```
+
+When rendering, the parent chosen for `site-1234` will be `region-1234`,
+since it is the highest precedence document that matches the label selector
+defined by `parentSelector`, and the parent chosen for `region-1234` will be
+`global-1234` for the same reason. The rendered result for `site-1234` would
+be:
+
+```yaml
+---
+schema: example/Kind/v1
+metadata:
+  name: site-1234
+data:
+  a:
+    z: 3
+  b: 4
+...
+```
+
+If `region-1234` were later removed, then the parent chosen for `site-1234`
+would become `global-1234`, and the rendered result would become:
+
+```yaml
+---
+schema: example/Kind/v1
+metadata:
+  name: site-1234
+data:
+  a:
+    x: 1
+    y: 2
+  b: 4
+...
+```
+
+<!-- TODO: Add figures for this example, with region present, have site point
+with dotted line at global and indicate in caption (or something) that it's
+selected for but ignored, because there's a higher-precedence layer to select
+-->
+
+### Substitution
+
+Substitution is primarily designed as a mechanism for inserting secrets into
+configuration documents, but works for unencrypted source documents as well.
+Substitution is applied at each layer after all merge actions occur.
```

Comment:
It seems like the interplay between substitution and layering should be clarified more.  Let me take a stab at some thoughts that I could try to include in the doc:

Layering order is only tangentially related to substitution.  The key constraints on substitutions from my perspective are:

1. Any part defined of any fully-rendered, concrete (not abstract) document is a valid substitution source.  The layer of that document is not a direct part of the substitution process.
2. When rendering a document, at each layer:
  1. Layering actions are applied
  2. Substitutions are applied

This combination provides a couple of properties:

- Due to the constraints on layering (can only select parents from lower precedence layers, can only have one parent), that portion of the compute graph cannot have cycles.
- Due to the constraint that we can only source substitutions from fully rendered documents, we likewise cannot have cycles when including them -- this will probably need detection during validation.

And enables a couple of nice features:

- We can have site-specific things like secrets added to documents which are supposed to be uniform for a region (like Armada definitions).
- The final document generated is deterministic because the order of all operations is unambiguous (though non-trivial).

This description is based on a really a naive, recursive conceptual understanding.  Of course the implementation can be e.g. fully iterative or graph based -- not trying to imply such an implementation constraint.

Are these thoughts helpful?